File: toggly-ruby/toggly/lib/toggly/telemetry/metrics_batcher.rb
# frozen_string_literal: true

module Toggly
  module Telemetry
    # Optional feature/variant correlation for a metric sample.
    MetricsFeatureOptions = Struct.new(:feature, :variant, keyword_init: true)

    # In-memory business metrics aggregator (Metrics.SendMetrics payload shape).
    class MetricsBatcher
      def initialize(app_key, environment, instance_name: nil)
        @app_key = app_key
        @environment = environment
        @instance_name = instance_name
        @measures = {}
        @counters = {}
        @observations = []
        @mutex = Mutex.new
      end

      def measure(metric, value, options = nil)
        @mutex.synchronize { add_to_map(@measures, metric, value.to_f, options) }
      end

      def increment_counter(metric, value = 1.0, options = nil)
        @mutex.synchronize { add_to_map(@counters, metric, value.to_f, options) }
      end

      def observe(metric, value, options = nil)
        @mutex.synchronize do
          opts = normalize_options(options)
          variant = opts&.variant && !opts.variant.to_s.empty? ? opts.variant.to_s : "enabled"
          feature = opts&.feature
          @observations << {
            time: Time.now.utc,
            metric: metric.to_s,
            feature: feature,
            variant: variant,
            value: value.to_f
          }
        end
      end

      def empty?
        @mutex.synchronize { @measures.empty? && @counters.empty? && @observations.empty? }
      end

      def build_and_reset
        @mutex.synchronize do
          return nil if @measures.empty? && @counters.empty? && @observations.empty?

          payload = {
            appKey: @app_key,
            environment: @environment,
            time: GrpcClients.to_protobuf_timestamp,
            stats: drain_map(@measures),
            counters: drain_map(@counters),
            observations: drain_observations
          }
          payload[:instanceName] = @instance_name if @instance_name
          payload
        end
      end

      def self.options_from_hash(options)
        return nil if options.nil?
        return options if options.is_a?(MetricsFeatureOptions)

        MetricsFeatureOptions.new(
          feature: options[:feature] || options["feature"],
          variant: options[:variant] || options["variant"]
        )
      end

      private

      def normalize_options(options)
        self.class.options_from_hash(options)
      end

      def metric_key(metric, feature)
        [metric.to_s, feature.to_s]
      end

      def add_to_map(store, metric, value, options)
        opts = normalize_options(options)
        variant = "enabled"
        feature = nil
        if opts
          variant = opts.variant.to_s unless opts.variant.nil? || opts.variant.to_s.empty?
          feature = opts.feature
        end
        key = metric_key(metric, feature)
        variants = store[key] ||= {}
        variants[variant] = variants.fetch(variant, 0.0) + value
      end

      def drain_map(store)
        out = []
        store.each do |(metric, feature), variants|
          variant_values = variants.reject { |_n, v| v.zero? }
          next if variant_values.empty?

          item = { metric: metric, variantValues: variant_values }
          item[:feature] = feature unless feature.empty?
          out << item
        end
        store.clear
        out
      end

      def drain_observations
        observation_messages = []
        groups = {}
        @observations.each do |obs|
          when_t = obs[:time]
          group_key = "#{when_t.to_f}\0#{obs[:metric]}\0#{obs[:feature] || ""}"
          group = groups[group_key]
          if group.nil? || group[:variantValues].key?(obs[:variant])
            group = new_observation_group(when_t, obs)
            groups[group_key] = group
            observation_messages << group
          end
          group[:variantValues][obs[:variant]] = obs[:value]
        end
        @observations = []
        observation_messages
      end

      def new_observation_group(when_t, obs)
        group = {
          time: GrpcClients.to_protobuf_timestamp(when_t),
          metric: obs[:metric],
          variantValues: {}
        }
        group[:feature] = obs[:feature] if obs[:feature] && !obs[:feature].to_s.empty?
        group
      end
    end
  end
end

File: Toggly.FeatureManagement.ReactRouter/react-router-toggly/src/core/telemetry/metrics-batcher.ts
import { toProtobufTimestamp } from './hash.js'

/** Max distinct metric keys (metric+feature) retained per measures/counters map. */
export const MAX_METRIC_KEYS_PER_BATCH = 500
/** Max observation rows retained before flush. */
export const MAX_OBSERVATIONS_PER_BATCH = 1_000

export interface MetricsFeatureOptions {
  /** Optional feature key for experiment-style subcounts. */
  feature?: string
  /** Variant name (default: enabled). */
  variant?: string
}

export interface MetricsBatcherOptions {
  appKey: string
  environment: string
  instanceName?: string
  maxMetricKeys?: number
  maxObservations?: number
}

type MetricKey = string // `${metric}\0${feature ?? ''}`

export interface MetricStatPayload {
  appKey: string
  environment: string
  time: { seconds: number; nanos: number }
  stats: Array<{
    metric: string
    feature?: string
    variantValues: Record<string, number>
  }>
  counters: Array<{
    metric: string
    feature?: string
    variantValues: Record<string, number>
  }>
  observations: Array<{
    time: { seconds: number; nanos: number }
    metric: string
    feature?: string
    variantValues: Record<string, number>
  }>
  instanceName?: string
}

/**
 * In-memory business metrics aggregator using variantValues maps (.NET parity).
 */
export class MetricsBatcher {
  private readonly appKey: string
  private readonly environment: string
  private readonly instanceName?: string
  private readonly maxMetricKeys: number
  private readonly maxObservations: number
  private measures = new Map<MetricKey, Map<string, number>>()
  private counters = new Map<MetricKey, Map<string, number>>()
  private observations: Array<{
    time: Date
    metric: string
    feature?: string
    variant: string
    value: number
  }> = []
  private droppedKeys = false

  constructor(options: MetricsBatcherOptions) {
    this.appKey = options.appKey
    this.environment = options.environment
    this.instanceName = options.instanceName
    this.maxMetricKeys = options.maxMetricKeys ?? MAX_METRIC_KEYS_PER_BATCH
    this.maxObservations = options.maxObservations ?? MAX_OBSERVATIONS_PER_BATCH
  }

  private key(metric: string, feature?: string): MetricKey {
    return `${metric}\0${feature ?? ''}`
  }

  private parseKey(key: MetricKey): { metric: string; feature?: string } {
    const sep = key.indexOf('\0')
    if (sep < 0) {
      return { metric: key }
    }
    const metric = key.slice(0, sep)
    const feature = key.slice(sep + 1)
    return feature ? { metric, feature } : { metric }
  }

  private addToMap(
    store: Map<MetricKey, Map<string, number>>,
    metric: string,
    value: number,
    options?: MetricsFeatureOptions,
  ): void {
    const variant = options?.variant ?? 'enabled'
    const mapKey = this.key(metric, options?.feature)
    let variants = store.get(mapKey)
    if (!variants) {
      if (store.size >= this.maxMetricKeys) {
        this.droppedKeys = true
        return
      }
      variants = new Map()
      store.set(mapKey, variants)
    }
    variants.set(variant, (variants.get(variant) ?? 0) + value)
  }

  measure(metric: string, value: number, options?: MetricsFeatureOptions): void {
    this.addToMap(this.measures, metric, value, options)
  }

  incrementCounter(metric: string, value = 1, options?: MetricsFeatureOptions): void {
    this.addToMap(this.counters, metric, value, options)
  }

  observe(metric: string, value: number, options?: MetricsFeatureOptions): void {
    if (this.observations.length >= this.maxObservations) {
      this.droppedKeys = true
      return
    }
    this.observations.push({
      time: new Date(),
      metric,
      feature: options?.feature,
      variant: options?.variant ?? 'enabled',
      value,
    })
  }

  isEmpty(): boolean {
    return this.measures.size === 0 && this.counters.size === 0 && this.observations.length === 0
  }

  hitCap(): boolean {
    return this.droppedKeys
  }

  private drainMap(
    store: Map<MetricKey, Map<string, number>>,
  ): Array<{ metric: string; feature?: string; variantValues: Record<string, number> }> {
    const out: Array<{ metric: string; feature?: string; variantValues: Record<string, number> }> =
      []
    for (const [key, variants] of store) {
      const parsed = this.parseKey(key)
      const variantValues: Record<string, number> = {}
      for (const [variant, value] of variants) {
        if (value !== 0) {
          variantValues[variant] = value
        }
      }
      if (Object.keys(variantValues).length > 0) {
        out.push({
          metric: parsed.metric,
          ...(parsed.feature ? { feature: parsed.feature } : {}),
          variantValues,
        })
      }
    }
    store.clear()
    return out
  }

  buildAndReset(): MetricStatPayload | null {
    if (this.isEmpty()) {
      return null
    }

    const observationGroups = new Map<
      string,
      {
        time: { seconds: number; nanos: number }
        metric: string
        feature?: string
        variantValues: Record<string, number>
      }
    >()

    for (const obs of this.observations) {
      const groupKey = `${obs.time.toISOString()}\0${obs.metric}\0${obs.feature ?? ''}`
      let group = observationGroups.get(groupKey)
      if (!group) {
        group = {
          time: toProtobufTimestamp(obs.time),
          metric: obs.metric,
          ...(obs.feature ? { feature: obs.feature } : {}),
          variantValues: {},
        }
        observationGroups.set(groupKey, group)
      }
      group.variantValues[obs.variant] = obs.value
    }
    this.observations = []
    this.droppedKeys = false

    const payload: MetricStatPayload = {
      appKey: this.appKey,
      environment: this.environment,
      time: toProtobufTimestamp(),
      stats: this.drainMap(this.measures),
      counters: this.drainMap(this.counters),
      observations: [...observationGroups.values()],
    }

    if (this.instanceName) {
      payload.instanceName = this.instanceName
    }

    return payload
  }

  restoreFromPayload(payload: MetricStatPayload): void {
    for (const stat of payload.stats ?? []) {
      for (const [variant, value] of Object.entries(stat.variantValues ?? {})) {
        this.measure(stat.metric, value, { feature: stat.feature, variant })
      }
    }
    for (const counter of payload.counters ?? []) {
      for (const [variant, value] of Object.entries(counter.variantValues ?? {})) {
        this.incrementCounter(counter.metric, value, {
          feature: counter.feature,
          variant,
        })
      }
    }
    for (const obs of payload.observations ?? []) {
      for (const [variant, value] of Object.entries(obs.variantValues ?? {})) {
        this.observe(obs.metric, value, { feature: obs.feature, variant })
      }
    }
  }
}
